mlir/lib/Dialect/Affine/Transforms/doc/AffineLoopInterchange.md

```diff
@@ -0,0 +1,136 @@
+## Affine Loop Interchange
+
+The file AffineLoopInterchange.cpp implements a loop interchange pass in the 
+affine dialect of MLIR. The pass deals only with rectangular loopnests which do 
+not have any affine.if statements. For all other cases, it simply bails out with-
+out crashing. The pass is driven by an analytical cost model that optimizes for 
+locality (spatial, temporal - both self and group) and parallelism for multicores
+to minimize the frequence of synchronization. It is able to handle both perfectly
+nested loop nests and imperfectly nested ones. The pass is triggered by using the
+command line flag -affine-loop-interchange.
+
+
+### Working:
+The pass starts by locating and converting all the imperfectly nested loopnests 
+to perfectly nested ones. This process is described in next section. Once all 
+such loop nests are converted, it starts dealing with each one of them in turn:
+
+*  First, it stores all the affine.for ops in a loopnest in a llvm::SmallVector.
+It assigns loop IDs to each induction variable of the AffineForOp it encounters. 
+The loop IDs are assigned in such a way that the outermost AffineForOp induction 
+variable has an ID of 0. These IDs are helpful in calculating the access matrices 
+for load/store ops.
+
+* Next, it iterates over all the possible permutations of a loopnest to determine 
+those permutations which do not violate any dependencies in the loop nest body. 
+Such permutations are called `validPermutations`. If only one valid permutation 
+is found, we do not need to proceed any further.
+
+* In case more than one permutations are valid, we may need to consider each of 
+those permutations to detect an optimal permutation. For this, we assign a cost to
+each permutation based on a measure of locality and the extent to which the loop 
+nest can be parallelized if this permutation is followed.We select the permutation 
+with the minimum cost as the best permutation.
+
+* The cost model needs few inputs to evaluate cost of each permutation - a list  
```

Comment:
evaluate cost -> evaluate the cost

Numerous errors like these.

Reply:
Fixed. Thanks.